README

                      _   _         ____
                     | | | | __ _  |  _ \ ___
                     | |_| |/ _` | | |_) / _ \
                     |  _  | (_| | |  _ <  __/
                     |_| |_|\__,_| |_| \_\___|

                    HaRe, the Haskell Refactorer
                a snapshot of our current prototype
                             07/07/2010
      http://www.cs.kent.ac.uk/projects/refactor-fp/hare.html

--------------------------------------------------------------------------------

This snapshot includes most dependencies (version A below), and
should build and work on Windows/Cygwin and Unix, with either gvim
6.2 or with emacs 21.4. 

HaRe_07072010 (HaRe 0.6) is still a prototype, made available so
you can play with basic refactoring support for Haskell, and give us
feedback or bug reports (see below for changes since HaRe 0.5 and
known remaining issues). The major new feature in this 0.6 release -
apart from bug fixes - is a number of new refactorings.
WE DO NOT RECOMMEND TO USE THIS PROTOTYPE ON YOUR PRODUCTION SOURCES JUST YET!

--------------------------------------------------------------------------------

--------------- what you need 

0 basics:

  1. ghc-6.12.1
  2. hint-0.3.2.3 (available from cabal)
  3. unix/gnu tools (or cygwin tools, if on windows)
  4. vim or emacs (we've tested with gvim 6.2 and with emacs 21.4)
  5. If you intend on running the test suite, HUnit-1.0.

A [recommended] partial build 
  (fewer tools needed, but tied to particular snapshot):

  1. fetch refactorer snapshot. this includes:
    - modified Programatica snapshot (includes happy output)
    - refactorer sources


--------------- one easy way to arrange these things

the following text assumes this directory structure
and unix/cygwin tools (adapt makefile if necessary):


  README.txt            <- you're reading this
  makefile
  editors               <- editor interface scripts for Vim/Emacs
  refactorer            <- refactorer 

  StrategyLib-4.0-beta  <- Strafunski's library for strategic programming
  diffs                 <- modified files for tools
  tools                 <- programatica snapshot

You might want to check the paths in makefile,
refactorer/HuMakefile and refactorer/myghc--make for sanity wrt to
your system (you should not actually need Happy for A). They should
work without change on windows/cygwin (windows/mingw has also been
reported to work), and feedback from the first release suggests the
same for unixish systems, from Max OS X to Linux (please let us know
if not, or if you can add new platforms). 

If on Suns, the call to ghc in refactorer/myghc--make needs a little
looking after: in principle, you could remove the SunOS*) case and
use the default case *). In practice, at least on Suns, you will
want to make sure that ghc uses the linker from binutils instead of
the Sun linker (order of magnitude difference in linking time..),
which is what setting GCC_EXEC_PREFIX before calling ghc does - just
put in the path to your local binutils installation.

***IMPORTANT*** for Mac OS users: please make sure that the .gvimrc
file contains the lines
 
          set noguipty
          set nocompatible
  

--------------- how to build   

  1. run: 
          runhaskell Setup.hs configure --prefix /path/to/hare
          (where /path/to/hare is a folder with read/write access)
          runhaskell Setup.hs build
          runhaskell Setup.hs install

          export PATH=/path/to/hare/bin:$PATH
          
--------------- how to use

  2. load the scripted interface into your favourite editor
     (you may or may not want to include this in your standard
      startup files for Haskell)

    - emacs: 
      M-x load-file /path/to/hare/share/hare-0.6/haskell-refac.el
      M-x haskell-refac-mode
      (to switch off: M-x haskell-refac-mode)
      You will then need to customize HaRe to point to your installation
      so Programatica can find its chase files.
      In the refactoring menu, go to customize -> chasepaths.
      Replace HaskellLibraries with:  /path/to/hare/share/hare-0.6
      Choose Save.

    - gvim:
      :source editors/refactor.vim
      :RefacStart	(or select start from menu)
      :RefacStop	(or select stop from menu)
      You will then need to customize HaRe to point to your installation
      so Programatica can find its chase files.
      In the refactoring menu, go to customize -> chasepaths.
      Replace HaskellLibraries with:  /path/to/hare/share/hare-0.6
      Choose Save.

  3. this should give you a menu with refactoring commands,
     and a separate refactoring process.

  4. the Programatica frontend we use is project-based.  before
     doing any refactorings, you have to start a project (select
     "new" to start a project with only the current module in it,
     then select "chase" to add the Prelude modules). project
     settings are kept in a "hi" directory, created in the current
     directory (it is recommended to edit files from within the
     directory they reside in).  

     ***IMPORTANT:*** Please re-create your projects once you have 
     updated your HaRe to a new snapshot/release.

  5. for most refactorings, the editor interface will pass the
     current filename (should be the module name, as well), the
     refactoring command, and the current cursor position. For
     some refactorings, you'll also need to highlight an 
     expression, or enter a new name. Here's the current list:
     
     Introduce Pattern Match    : place cursor over pattern variable
     
     Introduce Sub pattern      : place cursor over pattern variable
     
     Introduce Case Expression  : place cursor over pattern variable
     
     Generative Fold            : highlight expression to fold. 
                                  Must have equation in comment directly before
                                  definition of highlighted expression.
     
     Clone Analysis             : select from menu
     
     Clone Extraction           : highlight clone from the clone class to extract, follow
                                  step-by-step instructions.
     
     Converting Let to Where    : place cursor at start of identifer whose
                                  definition is to be moved
     
     Converting Where to Let    : place cursor at start of identifer whose
                                  definition is to be moved.
     
     create type signatures     : select from menu
     
     remove redundant declarations
                                : highlight definition to clean up
     
     add constructor to data
    type                        : place cursor at start of data type, you'll
                                  be prompted to add the new Constructor
                                  name, followed by any Type parameters as a 
                                  complete string
     Split a tuple              : place cureso at start of top level definition that
                                  returns a tuple. 
                                  
     Slicing based on a subexpression
                                : highlight sub-expression to extract.
                                
     Instantiate a general pattern
                                : select a function equation and enter 
                                  the new instances for each pattern in the argument set.
                                  
     
    
     Extract an expression      : highlight an expression within a definition
     Convert data type to
     newtype                    : place cursor at start of data type
     
     add definition to merge    : place cursor at start of definition that is to be merged
     
     merge definitions          : use "add definition for merge" for each definition to be merged.
                                  then choose merge definitions from menu.
                                  
     fold function definition   : highlight function equation to fold against.
     
     fold constant definition   : highlight constant equation to foldl against.
     
     convert pattern to an as pattern
                                : highlight pattern to convert.
    
     unfold references to as patterns
                                : place cursor over as pattern name.
                                
     remove field from a data type
                                : place cursor at start of field to remove.
                                
     add field to data type
                                : place cursor over start of constructor.
                                  enter new field name.
                                  
     
                                    
     add debug information      : add trace calls to function by placing cursor at the
                                  start of the function to trace
     

     rename                     : place cursor at start of identifier to
                                  be renamed, you'll be prompted for a
                                  new name

     Lift def to top level      : place cursor at start of identifier 
                                  whose definition is to be moved
     Lift def one level         : place cursor at start of identifier 
                                  whose definition is to be moved
     Demote                     : place cursor at start of identifier 
                                  whose definition is to be moved
     Introduce new def          : highlight expression to be named,
                                  you'll be prompted for a new name
     Unfold def                 : place cursor at start of identifier 
                                  where its definition is to be unfolded
     Generalise def             : place cursor at start of identifier 
                                  where its definition is to be unfolded
     Remove def                 : place cursor at start of identifier 
                                  whose definition is to be removed
     Duplicate def              : place cursor at start of identifier 
                                  whose definition is to be duplicated,
                                  you'll be prompted for a new name

     Add one parameter          : place cursor at start of identifier 
                                  whose definition is to be modified,
                                  you'll be prompted for a new name
     rm one parameter           : place cursor in formal parameter 
                                  which is to be removed
   
     Move def to another module : place cursor at start of the identifier
                                  whose definition is to be moved, you'll
                                  be prompted for the module name
     Clean imports              : cursor position does not matter, just 
                                  choose the command from the refactor menu
     Make import explicit       : place cursor at the start of the module 
                                  name in the import declaration
     Add to export              : place cursor at start of the identifier
                                  which is to be added to the export
     Remove from export         : place cursor at start of the entry to 
                                  be removed in the export list
     From concrete to abstract data type: place cursor at start of the data type name
     Add field names            : place cursor at start of the data type name
     Add discriminators         : place cursor at start of the data type name
     Add constructors           : place cursor at start of the data type name
     Eliminate nested patterns  : place cursor at start of the data type name
     Eliminate patterns         : place cursor at start of the data type name
     Create an ADT module       : place cursor at start of the data type name
     Eliminate intermidiate list: cusor position does not matter, just choose 
                                  the command from the menu.
6. If the 'Customize' menu in Emacs does not work, you can do the customisation by:
   M-x customize-group / haskell-refac


--------------- known issues

- the editor interface still needs improvement, although
  all known bugs have been fixed (windows vim users: it is 
  recommended to install vim in its standard location). 

- there is no standard for associating comments with
  semantic entities. we use heuristics to decide when
  to move/remove comments together with definitions
  => these heuristics may not match everyone's 
     commenting style

  outline: 

    when (re-)moving a definition, we also (re-)move the last
    consecutive comments right before this definition (with at most one
    empty line between the comments and the definition); we also
    (re-)move the comment whose start location is in the same line as
    the definition's last line of code. 

  Example:

    ....
    --Comment1

    --Comment2
    --Comment3

    fun1 x =  x+1  --Comment4
    --Comment5 

    ...
     
  In this case, Comment2, Comment3 and Comment4 belong to function fun1.

- if you're routinely working in a mixed Windows/Unix setup, you'll
  notice that the file-format (encoding of line endings) of HaRe
  output depends on whether you're using a Windows or a Unix HaRe, 
  not on whether the input is in Windows or Unix format.

- the Programatica frontend is currently limited to Haskell 98

- the Programatica frontend does not work with no-source files.

---------------change from 08/05/2009
-- general
   + make HaRe compatible with the latest version of GHC
   + ported to hint library, as an abstraction for GHC API.
   + bug fixes.

---------------change from 22/11/2006
-- general
   + make HaRE compatible with the latest version of GHC
   + several refactorings and bug fixes relating to slicing refactorings have 
     been added
   + API has been extended
   + Now uses GHC API for some refactorings 

---------------changes from 15/04/2005
-- general
   + make HaRe compatable with the latest version of GHC and Programatica.
   + Several refactorings related to slicing and data types have been 
     added.   
   + A pointwise to pointfree transformation has been added.

---------------changes from 11/03/2005
-- bugfix
   -- Introduce new defintion: only replaced the highlighted expression if the expression
      contains negative numbers.
   -- checking for inscoped identifiers.
   -- checking for conflicting export list.  
---------------changes from 03/12/2004
--gereral
  removed the unnecessary files in the snapshot.

--bugfix
   -- Introduce new definition: only replaced the highlighted expression if the expression
      contains literals.
----------------changes from 19/11/2004

-- general
  
   + Updated to a new Programatica version which supports hierarchical modules.
  
   + Adding an refactoring to the editor now only need to modify only file.

-- bugfixes
   + the 'Customize' menu did not work.   

----------------changes from 16/09/2004

-- general

  + The API. An API for program analysis and tranformation. Together with Programatica and Strafunski,
    this API can be used for the implementation for elementary refactorings. The file that contains 
    this API is RefacUtils.hs (the module name is RefacUtils). You can find the documentation for the
    API at HaRe_20012008/refactorer/doc. 
    
  + The hidden token stream manipulation. The basic idea is to move the token stream manipulations
    from the implementation of individual refactorings to the API, so that a program transformation
    function from the API modifies not only the AST but also the token stream. This liberates the 
    implementator of refactoring from caring about the token stream. 

  + For those users who are interested in the implementation of refactorings, we have written and 
    documented the implementation of two very simple refactorings: swapping the first two arguments
    of a function, and from it-then-else expression to case expression. You can find the implementation
    at HaRe_20012008/refactorer/RefacSwapArgs.hs, and HaRe_20012008/refactorer/RefacCase.hs.

  + We have a test suite to HaRe 0.3, and you can find it at HaRe_20012008/testing. There is a README file
    to explain how it works.


--------------- changes from  09/08/2004   

-- bugfixes
   + demote directly recursive definition.
   + non-break space

---------------changes from 26/04/2004
- features
   + The refactor menu has been switched to hierarchical menus to reduce clutter.
   + In 'generalise a definition', 'replace all' has been switched on. 

- refactoring
   - new refactorings:
    + 'from concrete to abstract data type': transforms a concrete data type defintion into an abstract one.
      This is a composite refactoring. It consists of five elementary refactorings:
      + 'Add field names': add field names to a data type declaration.
      + 'Add disriminators': add a discriminator for each data constructor defined in the identified data type.
      + 'Add constructors': add a constructor function for each data constructor defined in the identified data type.
      + 'Eliminate nested patterns': remove patterns nested in a pattern specified by the data constructor defined
                                     in the identified data type.
      + 'Eliminate patterns': remove those patterns specified by the data constructor defined in the identified data type.
      + 'Create an ADT module': create the ADT module interface.
    + 'eliminate intermidiate list' (this refactoring is still in an experimental stage)

   -bug fixes
    + After generalising a function with a type signature, the type signature is now commented out in the program source.
    + wrong layout when refactoring on operator functions.
    + wrong layout when refactoring on an expression ending with [] or ()

--------------- changes from 26/02/2004

- general 

   - update to a new Programmatica version  

- refactoring

   - new refactorings:
     + 'move def to another module' : move a definition from the current
                                      module to a user-specified module
     + 'clean imports'        : remove redundancies from the imports 
     + 'make import implicit' : make the used entities from an
                                import explicit in the declaration
     + 'add to export'        : add an entity to the export list
     + 'remove from export'   : remove an entity from the export list
   
   - bugfixes:

    + qualified uses of a top-level identifier becomes 
      unqualified after its definition has been demoted. 

--------------- changes from 27/01/2004

- refactoring

  - features:
    + in 'introduce a new definition', 'replace all' has been
      switched on. The refactorer should be able to detect all
      identical expressions in the scope.
    + operators can now be renamed

  - bugfixes:
    + module names can differ from file names
    + avoid dangling parenthesis after unfoldDef 
      (could trigger layout rule syntax errors)
    + do not reverse names of new variables introduced
      by unfoldDef (did you know that 0_x is legal?-)
    + do not trip over hard tabs
    + give an error message when trying to lifting a definition 
      in a class/instance declaration to the top level.
    + only write client modules when they are really changed.
    + renaming data types should not rename data constructors as well
    + lifting defs should not add globally available ids as parameters
    + the reported layout bugs with lifitng and intro new defs have
      been fixed

  - known issues
    + hard tabs are not counted correctly yet
    + refactorings shouldn't introduce hard tabs!

- editor interfaces

  - features:
    + Emacs: thanks to help from JP Bernardy, we have now moved the
      emacs lisp script to a subset that both Emacs and XEmacs can 
      work with; in the process, the gui has been cleaned up a bit
      and should now operate more like a normal minor mode.. 
    + Vim: don't try to open files modified by the refactorer,
      unless they are already associated with a buffer; show
      warning dialog instead (now in line with emacs version)

  - bugfixes:
    + Emacs: don't create separate refactoring processes per buffer

  - known issues
    + in Vim, it is still possible to start refactorings while
      there are modified buffers (fixed in Emacs a while ago)

--------------- changes from 01/10/2003

- general:


ChangeLog
2017-06-12 v0.8.4.1
	* Update for ghc-mod 5.8.0.0
2017-02-11 v0.8.4.0
	* Update for GHC 8.0.2
	* Use ghc-mod version to 5.7
	* Add deleteDef refactoring to generalise a monadic function
	to use applicative (@SAdams601)
	* Add genapplicative refactoring to generalise a monadic function
	to use applicative (@SAdams601)
2016-08-03 v0.8.3.0
	* Update for GHC 8.0.1
	* Use ghc-mod version to 5.6 now it is released
2016-01-20 v0.8.2.3
	* Use ghc-mod version to 5.5 now it is released
	* Expose commands working in RefactGhc, so the ghc-mod session can
	be shared in haskell-ide-engine. By @cocreature
	* Internal only, heavily clean up the ...Refactoring.Rename
	module. By @glittershark
2016-01-07 v0.8.2.2
	* Lock ghc-mod version to 0.5.4 to prevent issues when 0.5.5 comes out
	* Set minimum bound on ghc-exactprint, the behaviour of layout has
	changed for ifToCase. This will be re-done once ghc-exactprint is
	updated for GHC 8.0.1
	* The constructors for ghc-mod `Options` are now re-exported too
	for convenience.
2015-11-08 v0.8.2.1
	* Set GHC-Opts to -O0 for sane compile times
	* Make the ghc-hare depend on the library, rather than rebuilding
	* elisp no longer prompts to save modified dired buffers
	all of it.
2015-10-07 v0.8.2.0
	* Work around a bug in GHC 7.10.2 that will not process pragmas
	after a comment at the top of a file, when comments are returned
	for ghc-exactprint processing.
	* Add parameters to all use sites in the original location when
	lifting a declaration.
2015-10-05 v0.8.1.1
	* Move to new/old home at https://github.com/RefactoringTools/HaRe
2015-10-05 v0.8.1.0
	* Reinstate processing of CPP directives as comments #36
	* Fix problem in renaming in a client module when the old name is
	not in scope in the client module. #35
	* No longer mark a client module as refactored if no renaming
	takes place in it.
2015-10-04 v0.8.0.0
	* Major rework, replacing haskell-token-utils with ghc-exactprint
	* Now supports GHC 7.10.2 and up. ONLY.
	* Functionally equivalent to v0.7.2.8
2014-10-13 v0.7.2.8
	lower bounds on parsec and haskell-token-utils
2014-08-20 v0.7.2.7
	Upper bound on ghc-mod until 5.0 is released and stabilises
2014-07-20 v0.7.2.6
	tokenise / basicTokenise moved to haskell-token-utils
2014-07-16 v0.7.2.5
	Fix incompatibility with haskell-token-utils 0.0.0.3, closing
	issue #14
2014-07-03 v0.7.2.4
	Re-architect to split out haskell-token-utils as a separate
	package, usable with haskell-src-exts too
2014-05-02 v0.7.2.3
	* Update for ghc-mod 4.1.0
2014-04-03 v0.7.2.2
	* Fix renaming so that it also renames types in the export list of
	a module
	* revert to ghc-mod >= 4.0, with note that FFI exports can't
	currently be processed
2014-04-01 v0.7.2.1
	* limit ghc-mod to <= 3.1.7, new version 4.0 does not deal with
	FFI exports. This does not stop ghc-mod >= 4.0 being used in the
	editor, it is just the library linked into ghc-hare
	* Added disclaimer that API is subject to change without notice
	* liftToTopLevel now brings along an updated type signature,
	provided there are no constraints on any of the types
2014-03-23 v0.7.2.0
	* Reorganised package structure to have
	Language.Haskell.Refact.HaRe which provides all the functions
	needed to call the refactorer from external code, and
	Language.Haskell.Refact.API which provides the interfaces to write
	new refactorings
	* liftToMod of last local decl having preceding comment now removes 'where'
	* Correctly determine the location for a demoted decl
2014-03-22 v0.7.1.7
	* Correctly determine what parameters to add when lifting a decl
	* Do not demote a decl if it is used in more than one match
2014-03-19 v0.7.1.6
	* Do not treat single-line layout as a group, hence preserving
	layout of adjacent let expressions when renaming.
	* Process Template Haskell expressions instead of blowing up
2014-03-17 v0.7.1.5
	* Adding parameters to lifted declarations now works as expected
2014-03-09 v0.7.1.4
	* Load and process files using TH and/or QuasiQuotes
	* Make loading of multiple targets from a cabal file more robust
2014-03-06 v0.7.1.3
	* Various fixes for robustness based on real world trials
	* Rework the free and declared variable calculation to be more accurate
	* Extend the range of syntax elements that HaRe can handle when
	reading a file.  Most bar parallel extensions are in place.
	* Make sure the refactored file retains its original extension, so
	that syntax highlighting still works while the preview is being done.
2014-01-28 v0.7.1.2
	* Ensure that the right target is activated for the primary file
	when it belongs to a non-library target,not just for client
	modules of it.
2014-01-27 v0.7.1.1
	* The non-library targets were having the wrong path prepended
2014-01-26 v0.7.1.0
	* Rework the module loading to load each target in turn, so that
	all exe, test, benchmark etc will be refactored in one go
2013-12-18 v0.7.0.9
	* Tweaks to output stage based on real world use. Renaming seems
	stable, liftXXX, demote have issues
	* Removed deprecated hspec-discover dependency
	* added ghc-hare show command to the elisp as well to be able to
	check that it is using the correct cabal file
2013-12-16 v0.7.0.8
	* Major rewrite of the token output stage. It now makes use of a
	dual-tree structure to manage the maintenance of required vertical
	layout while refactoring.
2013-10-21 v0.7.0.7
	* Bump the lower bound on Diff
2013-10-01 v0.7.0.6
	* Sort out most of the do/in/let layout when renaming. Only nested
	layout changes to be dealt with.
	* Updated to use the latest ghc-mod, with cabal 1.18.x sandbox
	support.
	* Updated to latest versions of Diff and hspec
	* When refactoring a CPP pre-processed file, read the preprocessor
	directives and non-compiled code as comments, so the source can be
	round-tripped. This allows refactoring of simple #if/#else/#endif
	code. NOTE: no refactoring is done on code that is not live after
	the preprocessor.
2013-09-12  v0.7.0.5
	* Now able to get tokens for a file pre-processed with CPP. But it

src/Main.hs
{-# LANGUAGE FlexibleInstances #-}
{-# OPTIONS_GHC -fno-warn-orphans #-}


import Control.Applicative       ( (<$>), (<*>) )
import Data.List
import Data.Maybe
import Exception
import System.Console.CmdTheLine
import qualified Text.PrettyPrint as PP

import Language.Haskell.GhcMod

-- The actual refactorings
import Language.Haskell.Refact.DupDef
import Language.Haskell.Refact.MoveDef
import Language.Haskell.Refact.Renaming

import Language.Haskell.Refact.Utils.Monad
import Language.Haskell.Refact.Utils.TypeSyn
import qualified Language.Haskell.Refact.Case as GhcRefacCase

-- Based initially on http://elifrey.com/2012/07/23/CmdTheLine-Tutorial/

-- ---------------------------------------------------------------------

instance ArgVal (Maybe RefactSettings) where
   converter = just

instance ArgVal (Maybe SimpPos) where
  converter = just

instance ArgVal RefactSettings where
  converter = (pRefactSettings,ppRefactSettings)
    where
        pRefactSettings :: ArgParser RefactSettings
        pRefactSettings _s = Right (RefSet ["."] Normal)

        ppRefactSettings :: ArgPrinter RefactSettings
        ppRefactSettings s = PP.text (show s)


instance ArgVal SimpPos where
  converter = (pSimpPos, ppSimpPos)
    where
        pSimpPos :: ArgParser SimpPos
        pSimpPos s = res
          where
            res = case ((fmap fst . listToMaybe . reads) s) of
              Nothing -> Left $ PP.text "Expecting (row,col)"
              Just r -> Right r

        ppSimpPos :: ArgPrinter SimpPos
        -- ppSimpPos = error "ppSimpPos undefined"
        ppSimpPos s = PP.text (show s)


-- ---------------------------------------------------------------------
-- command line argument types

-- |cradle from ghc-mod
{-
cradle :: Term (Maybe Cradle)
cradle = do
  (strVer,ver) <- getGHCVersion
  return undefined
-}

-- As used in ghc-mod
--    (strVer,ver) <- getGHCVersion
--    cradle <- findCradle (sandbox opt') strVer



mainFile :: Term (Maybe FilePath)
mainFile = value $ opt Nothing (optInfo ["m","main"])
     { optName = "main"
     , optDoc  = "The Haskell main file, if there is one"
     , optSec = comOpts
     }

settings :: Term (Maybe RefactSettings)
settings = value $ opt (Just defaultSettings) (optInfo [ "s", "settings"])
     { optName = "settings"
     , optDoc  = "The set of paths to the source"
     , optSec = comOpts
     }


thisFile :: Term FilePath
thisFile = required $ pos 0 Nothing posInfo
    { posName = "file"
    , posDoc = "The file originating the refactoring."
    }

startRow :: Int -> Term Int
startRow p = required $ pos p Nothing posInfo
    { posName = "startline"
    , posDoc = "The line of the start of the refactoring."
    }

startCol :: Int -> Term Int
startCol p = required $ pos p Nothing posInfo
    { posName = "startcol"
    , posDoc = "The col of the start of the refactoring."
    }

endRow :: Int -> Term Int
endRow p = required $ pos p Nothing posInfo
    { posName = "endline"
    , posDoc = "The line of the end of the refactoring."
    }

endCol :: Int -> Term Int
endCol p = required $ pos p Nothing posInfo
    { posName = "endcol"
    , posDoc = "The col of the end of the refactoring."
    }

newName :: Int -> Term String
newName p = required $ pos p Nothing posInfo
    { posName = "newname"
    , posDoc = "The new name."
    }


-- ---------------------------------------------------------------------

-- 'input' is a common option. We set its 'optSec' field to 'comOpts' so
-- that it is placed under that section instead of the default '"OPTIONS"'
-- section, which we will reserve for command-specific options.
input :: Term (Maybe String)
input = value $ opt Nothing (optInfo [ "input", "i" ])
      { optName = "INPUT"
      , optDoc = "For specifying input on the command line. If present, "
               ++ "input is not read form standard-in."
      , optSec = comOpts
      }


-- ---------------------------------------------------------------------

main :: IO ()
main = runChoice defaultTerm [ ifToCaseTerm
                             , duplicateDefTerm
                             , liftToTopTerm
                             , liftOneTerm
                             , demoteTerm
                             , renameTerm
                             ]

defaultTerm :: (Term a, TermInfo)
defaultTerm = ( ret $ const (helpFail Pager Nothing) <$> mainFile -- <*> settings
              , info
              )
  where
    info :: TermInfo
    info = defTI'
      { termName = "ghc-hare"
      , version  = "0.7.0.0"
      , termDoc  = doc
      }


ifToCaseTerm :: (Term (IO ()), TermInfo)
ifToCaseTerm = (doIfToCase, info)
  where

    doIfToCase :: Term (IO ())
    doIfToCase = ifToCase <$> settings <*> mainFile <*> thisFile
               <*> startRow 1 <*> startCol 2 <*> endRow 3 <*> endCol 4

    ifToCase s m f sr sc er ec = runFunc $ GhcRefacCase.ifToCase s m f (sr,sc) (er,ec)

    info :: TermInfo
    info = defTI'
      { termName = "iftocase"
      , termDoc  = "Convert an if expression to a case expression"
      }

-- ---------------------------------------------------------------------

-- duplicateDef :: Maybe RefactSettings -> Maybe FilePath -> FilePath -> String -> SimpPos -> IO ()
-- duplicateDef settings maybeMainFile fileName newName (row,col) =

duplicateDefTerm :: (Term (IO ()), TermInfo)
duplicateDefTerm = (doDupDef,info)
  where
    doDupDef :: Term (IO ())
    doDupDef = dupDef <$> settings <*> mainFile <*> thisFile <*> newName 1 <*> startRow 2 <*> startCol 3

    dupDef s m f nn r c = runFunc $duplicateDef s m f nn (r,c)

    info :: TermInfo
    info = defTI'
      { termName = "dupdef"
      , termDoc  = "Duplicate a definition"
      }

-- ---------------------------------------------------------------------

-- liftToTopLevel :: Maybe RefactSettings -> Maybe FilePath -> FilePath -> SimpPos -> IO ()
-- liftToTopLevel settings maybeMainFile fileName (row,col) =

liftToTopTerm :: (Term (IO ()), TermInfo)
liftToTopTerm = (doLiftToTop,info)
  where
    doLiftToTop :: Term (IO ())
    doLiftToTop = liftToTop <$> settings <*> mainFile <*> thisFile <*> startRow 1 <*> startCol 2

    liftToTop s m f r c = runFunc $ liftToTopLevel s m f (r,c)

    info :: TermInfo
    info = defTI'
      { termName = "liftToTopLevel"
      , termDoc  = "Lift a declaration to the top level"
      }

-- ---------------------------------------------------------------------

-- liftOneLevel :: Maybe RefactSettings -> Maybe FilePath -> FilePath -> SimpPos -> IO ()
-- liftOneLevel settings maybeMainFile fileName (row,col) =

liftOneTerm :: (Term (IO ()), TermInfo)
liftOneTerm = (doLiftOne,info)
  where
    doLiftOne :: Term (IO ())
    doLiftOne = liftOne <$> settings <*> mainFile <*> thisFile <*> startRow 1 <*> startCol 2

    liftOne s m f r c = runFunc $ liftOneLevel s m f (r,c)

    info :: TermInfo
    info = defTI'
      { termName = "liftOneLevel"
      , termDoc  = "Lift a declaration one level"
      }

-- ---------------------------------------------------------------------

-- demote :: Maybe RefactSettings -> Maybe FilePath -> FilePath -> SimpPos -> IO ()
-- demote settings maybeMainFile fileName (row,col) =

demoteTerm :: (Term (IO ()), TermInfo)
demoteTerm = (doDemote',info)
  where
    doDemote' :: Term (IO ())
    doDemote' = demote' <$> settings <*> mainFile <*> thisFile <*> startRow 1 <*> startCol 2

    demote' s m f r c = runFunc $ demote s m f (r,c)

    info :: TermInfo
    info = defTI'
      { termName = "demote"
      , termDoc  = "Move a declaration down one level"
      }

-- ---------------------------------------------------------------------

-- rename :: Maybe RefactSettings -> Maybe FilePath -> FilePath -> String -> SimpPos -> IO [FilePath]
-- rename settings maybeMainFile fileName newName (row,col) =

renameTerm :: (Term (IO ()), TermInfo)
renameTerm = (doRename,info)
  where
    doRename :: Term (IO ())
    doRename = rename' <$> settings <*> mainFile <*> thisFile <*> newName 1 <*> startRow 2 <*> startCol 3

    rename' s m f n r c = runFunc $ rename s m f n (r,c)

    info :: TermInfo
    info = defTI'
      { termName = "rename"
      , termDoc  = "Rename an identifier"
      }

-- ---------------------------------------------------------------------

runFunc :: IO [String] -> IO ()
runFunc f = do
  r <- catchException f
  let ret = case r of
       Left s    -> "(error " ++ (show s) ++ ")"
       Right mfs -> "(ok " ++ showLisp mfs ++ ")"
  putStrLn ret


showLisp :: [String] -> String
showLisp xs = "(" ++ (intercalate " " $ map show xs) ++ ")"

catchException :: (IO t) -> IO (Either String t)
catchException f = do
  res <- handle handler (f >>= \r -> return $ Right r)
  return res
  where
    handler:: SomeException -> IO (Either String t)
    handler e = return (Left (show e))

-- ---------------------------------------------------------------------

doc :: String
doc = "Haskell Refactorer"


-- The heading under which to place common options.
comOpts :: String
comOpts = "COMMON OPTIONS"

-- A modified default 'TermInfo' to be shared by commands.
defTI' :: TermInfo
defTI' = defTI
  { man =
      [ S comOpts
      , P "These options are common to all commands."
      , S "MORE HELP"
      , P "Use '$(mname) $(i,COMMAND) --help' for help on a single command."
      , S "BUGS"
      , P "Email bug reports to <foo@bar.com>"
      ]
  , stdOptSec = comOpts
  }

